cql3/functions/aggregate_fcts.cc

```diff
@@ -183,15 +225,62 @@ class impl_avg_function_for final : public aggregate_function::aggregate {
         ++_count;
         _sum += value_cast<Type>(data_type_for<Type>()->deserialize(*values[0]));
     }
+    virtual void set_accumultor(const opt_bytes& acc) override {
+        /*data_type tuple_type = tuple_type_impl::get_instance({data_type_for<Type>, long_type});
+        auto tuple = value_cast<tuple_type_impl::native_type>(tuple_type->deserialize(bytes_view(*acc)));
+
+        _sum = value_cast<Type>(tuple[0]);
+        _count = value_cast<int64_t>(tuple[1]);*/
+    }
```

Comment:
See apache/cassandra@7872318d63009193415ba1365beedc2303a92386. If we make the accumulator an ABI, we have to be careful with how it is implemented, since mistakes are now more costly to fix.

Reply:
Could you specify this comment a little more? Would you like to implement AVG state like Cassandra (pair of current avg and counter)?

Reply:
There are two questions:
1. should we switch to Cassandra's implementation?
2. if so, when should we switch? before or after this series?

The answer to the first question if likely yes. Quoting the JIRA:

> For fixed-size integer types overflow is all but guaranteed to happen, yielding incorrect result. While for sum it is somewhat acceptable as the result cannot fit the type, this is not the case for average.

We don't suffer from that because we use a wide accumulator. We should probably keep our current implementation, but if we change the floating-point implementation, we can export and import the accumulator as average+count.

> As the result of average is always within the scope of the source type, failing to produce it only signifies a bad implementation. Yes, one can solve this by type-casting, but do we really want to always have to be telling people that the correct spelling of the average function is cast(avg(cast(value as bigint))) as int), especially if this is so trivial to fix?
>
> Additionally, the straightforward addition we use for floating point versions is not a good choice numerically for larger numbers of values. We should switch to a more stable version, e.g. iterative mean using avg = avg + (value - avg) / count.

We should look for evidence that this is correct.

If we do change, we should change before this series is merged, otherwise making the change later involves a cluster feature and tagging every RPC to indicate which version of the accumulator we use.

Maybe a simple change is to just change the accumulator import/export format to match Cassandra. Then we're insulated from the effects of the change. We also don't have to worry about using varint.

Reply:
We've been discussing with @psarna and our conclusion was we probably should stay with sum+count state. 
In the reducible AVG we still need to have sum to reduce two state into one, so in avg+count model this requires multiplying and using wide accumulator to not overflow the sum.

So to really benefit from avg+count implementation the aggregation shouldn't be distributed.

Reply:
> We've been discussing with @psarna and our conclusion was we probably should stay with sum+count state. In the reducible AVG we still need to have sum to reduce two state into one, so in avg+count model this requires multiplying and using wide accumulator to not overflow the sum.

It's no longer an accumulator but a temporary variable. I think it's better not to have to transport varints around.

> 
> So to really benefit from avg+count implementation the aggregation shouldn't be distributed.

What about the the other point about floating-point stability?

Reply:
Why transporting varint is a problem? Is it good enough argument to switch to avg+count state only to not transport varint?

Isn't floating-point stability argument also breaking in our case where we have to calculate by multiplying avg * count?

Reply:
We can avoid varint  but the intermediate avg has to be big_decimal and transported as `decimal_type`.

So following the sync conversation, should we switch to avg+count?

Reply:
Let's keep the current implementation. I think it's compatible with the more involved algorithm (we just have to multiply the current average with the count before transporting it).